torch/fx/passes/infra/partitioner.py

```diff
@@ -73,33 +73,40 @@ def maybe_merge_partition(self_id: int, other_id: int):
             # the set.
             visited: Set[Node] = set()
 
-            def dfs_find_cycle(node):
-                if node in visited:
-                    return False
-                if node in merged_nodes:
-                    return True  # found cycle, return
-
-                visited.add(node)
-                # branching on hitting partition or not
-                if node in assignment:
-                    # Since partition is not merged in the graph yet, when we
-                    # hit a node in a partition through DFS, we need to
-                    # traverse all nodes in the partition to properly reflect
-                    # dependencies after the fusion
-                    for p_node in partitions_by_id[assignment[node]].nodes:
-                        for user_node in p_node.users:
-                            if user_node not in partitions_by_id[assignment[node]].nodes and dfs_find_cycle(user_node):
-                                return True
-                else:
-                    for user_node in node.users:
-                        if dfs_find_cycle(user_node):
-                            return True
+            def iter_find_cycle(root_node):
+                stack : List[Node] = []
+                stack.append(root_node)
+
+                while len(stack) != 0:
+                    node = stack.pop(0)
```

Comment:
Because `stack` is a Python list, popping from the beginning is O(n). 
Also it looks like adding to the end on line 100 and popping from the beginning here make the whole thing BFS instead of DFS.

Reply:
Hahaha, looks like I failed the coding interview :laughing: 

Wasn't paying much attention to BFS vs DFS. I think in the context of cyclic dependency, DFS does make more sense. I'll update it.